src/main/webapp/view/discussion/discussion-manager.js

```diff
@@ -24,57 +24,66 @@ export default class DiscussionManager {
   static #PARAM_TIMESTAMP = 'timestamp';
   static #PARAM_TYPE = 'type';
   #lecture;
+  #displayedComments;
 
   /**
    * Creates a `DiscussionManager` to manage posting and fetching comments for
    * the `lecture`.
    */
   constructor(lecture) {
     this.#lecture = lecture;
+    this.#displayedComments = new Map();
   }
 
   /**
-   * Organizes comments into threads with nested replies.
-   *
-   * <p>All comments are sent from the servlet without any structure, so the
-   * client needs to organize them before displaying.
+   * Adds new comments to threads with nested replies by setting comment's
+   * parent and replies fields. Only processes and returns comments with ids
+   * that haven't already been seen. This is a private method and should only
+   * be called by `DiscussionManager`.
    */
-  structureComments(allComments) {
-    const commentIds = {};
+  processNewComments(allComments) {
+    const newComments = [];
+
+    // Comments are not guarenteed to have any specific order, so we need to do
+    // this loop before the other one to filter new comments and initialze
+    // fields.
     for (const comment of allComments) {
+      const id = comment.commentKey.id;
+      if (this.#displayedComments.has(id)) {
+        // This comment is already displayed.
+        continue;
+      }
       comment.replies = [];
-      commentIds[comment.commentKey.id] = comment;
+      newComments.push(comment);
+      this.#displayedComments.set(id, comment);
     }
 
-    const rootComments = [];
-    for (const comment of allComments) {
+    // Sets parent element, and adds comments as replies, since all are
+    // initialized in first loop.
+    for (const comment of newComments) {
       if (comment.type === COMMENT_TYPE_REPLY) {
-        const parent = commentIds[comment.parentKey.value.id];
+        const parentId = comment.parentKey.value.id;
+        const parent = this.#displayedComments.get(parentId);
         parent.replies.push(comment);
-      } else {
-        // Top level comments don't have parents.
-        rootComments.push(comment);
+        comment.parent = parent;
       }
     }
-    // Sort comments such that earliest timestamp is first.
-    rootComments.sort((a, b) => (a.timestampMs.value - b.timestampMs.value));
-    return rootComments;
+    return newComments;
   }
 
   /**
    * Fetches all of the lecture comments from the `ENDPOINT`. This
-   * returns a sorted array of all root comments, with replies added to
-   * their parents.
+   * returns an array of all comments that haven't been fetched before.
```

Comment:
nit: Fetches and returns all comments that have not been fetched before from `ENDPOINT`.

Reply:
done